### PR TITLE
Chore(repo): Migration to Yarn Modern

### PR DESCRIPTION
**Pull request by literat**
## Description

Yarn v1 is an old implementation and has been recommended to migrate to modern versions. Today's version is v4 (https://yarnpkg.com/blog/release/4.0) - we cannot go to v4 since it requires Node.js v18.
We are migrating to the latest v3 (https://v3.yarnpkg.com/) which is far more performant and has better verbose messages.

There are a few changes like the configuration going through the `yarnrc.yml` or that the `pre` and `post` hooks do not work (due to possible collisions in the script lifecycle).

Please, give this everyone a try. I need to know that everything is working for you. And finally we will be a little more up-to-date :-)

### Additional context

  * @see: https://yarnpkg.com/getting-started/install

There were some additional modifications that need to be done:
  * adding missing dependencies
  * fixing Stylelint errors - @adamkudrna and @crishpeen please, have a look
  * fixing build scripts

The changes are in separate commits to allow better fixing.

### Issue reference

refs #264

---

### Before submitting the PR, please make sure you do the following

- [ ] Read the [Contributing Guidelines](https://github.com/lmc-eu/spirit-design-system/blob/main/CONTRIBUTING.md).
- [ ] Follow the [PR Title/Commit Message Convention](https://github.com/lmc-eu/spirit-design-system/blob/main/CONTRIBUTING.md#commit-conventions).
- [ ] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

### CODE REVIEW

**Review by pavelklibani — APPROVED** [review]

**Comment by literat** [comment]
https://github.com/jupyter-lsp/jupyterlab-lsp/pull/1021

**Comment by literat** [comment]
https://yarnpkg.com/features/linkers

File: package.json [hunk 1]
```diff
@@ -109,5 +108,6 @@
   "resolutions": {
     "@types/react": "18.2.37",
     "typescript": "4.7.4"
-  }
+  },
+  "packageManager": "yarn@3.7.0"
```
Comment (adamkudrna) [inline package.json:112]: This is an unofficial property I guess…?

https://docs.npmjs.com/cli/v10/configuring-npm/package-json

File: .yarnrc.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+nodeLinker: node-modules
+
+yarnPath: .yarn/releases/yarn-3.7.0.cjs
```
Comment (adamkudrna) [inline .yarnrc.yml:3]: Why a local instance? `npm install yarn -g` is not a way to go?

File: .stylelintrc.js [hunk 3]
```diff
@@ -7,5 +7,6 @@ module.exports = {
     // This rule can lead to less readable code
     // @see: https://github.com/stylelint/stylelint-config-recommended/issues/14
     'declaration-block-no-redundant-longhand-properties': null,
+    'color-hex-length': 'long',
```
Comment (adamkudrna) [inline .stylelintrc.js:10]: Since all colors are in genereated code (and we plan a revision of our Stylelint config anyway), I'm OK with this rule.

⚠️  **Just remember we must update our Sass exporter so it produces the colors in correct format.**

cc @crishpeen

File: .gitignore [hunk 4]
```diff
@@ -38,3 +38,13 @@ packages/*/build
 # This prevents accidental commit of this token to the repository.
 nx-cloud.env
 .nx
+
+# Yarn Modern
+# @see: https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored
+.pnp.*
```
Comment (adamkudrna) [inline .gitignore:44]: Is this also related to Yarn?

File: packages/analytics/package.json [hunk 5]
```diff
@@ -51,6 +52,7 @@
     "@typescript-eslint/eslint-plugin": "6.18.0",
     "@typescript-eslint/parser": "6.18.0",
     "jest": "29.7.0",
+    "npm-run-all": "^4.1.5",
```
Comment (adamkudrna) [inline packages/analytics/package.json:55]: +1 for using it in multiple packages.

File: packages/web-react/config/vite/app.ts [hunk 6]
```diff
@@ -47,6 +47,11 @@ export default defineConfig({
       },
     },
   },
+  resolve: {
+    alias: {
+      '@lmc-eu/spirit-icons': resolve(__dirname, '../../../icons/dist'),
```
Comment (adamkudrna) [inline packages/web-react/config/vite/app.ts:52]: 👍🏻

File: packages/web-react/src/components/Accordion/demo/index.tsx [hunk 7]
```diff
@@ -4,7 +4,7 @@ import React from 'react';
 import ReactDOM from 'react-dom/client';
 // eslint-disable-next-line @typescript-eslint/ban-ts-comment, import/extensions, import/no-unresolved
 // @ts-ignore: No declaration file
-import icons from '@lmc-eu/spirit-icons/dist/icons';
+import icons from '@lmc-eu/spirit-icons/icons';
```
Comment (adamkudrna) [inline packages/web-react/src/components/Accordion/demo/index.tsx:7]: 👍🏻

File: packages/web-react/src/components/Collapse/demo/index.tsx [hunk 8]
```diff
@@ -4,16 +4,16 @@ import React from 'react';
 import ReactDOM from 'react-dom/client';
 // eslint-disable-next-line @typescript-eslint/ban-ts-comment, import/extensions, import/no-unresolved
 // @ts-ignore: No declaration file
```
Comment (adamkudrna) [inline packages/web-react/src/components/Collapse/demo/index.tsx:6]: Is this comment still relevant? If not, we can happily delete it. If yes, a reason would be helpful.

File: .gitignore (hunk 4 above)
Comment (literat) [reply .gitignore:44]: I have used this setting from the documentation (https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored). Since we are not using zero-install we can safely ignore this.

File: packages/web-react/src/components/Collapse/demo/index.tsx (hunk 8 above)
Comment (literat) [reply packages/web-react/src/components/Collapse/demo/index.tsx:6]: This is related to the https://jira.almacareer.tech/browse/DS-561. Icons are not exporting the declaration files for Typescript usage.

Comment (literat) [reply packages/web-react/src/components/Collapse/demo/index.tsx]: I have modified the comment in all places. Thanks.

File: .stylelintrc.js (hunk 3 above)
Comment (literat) [reply .stylelintrc.js:10]: I have fixed this in https://github.com/lmc-eu/spirit-design-system/pull/1029/commits/c870602e7d62806844aed175751ef64e593ecede

Comment (crishpeen) [reply .stylelintrc.js:10]: I would:
1. remove this line
2. update exporter so it shortens even with alpha channel (#ffffff00 -> #fff0)
3. test exporter it returns short version everywhere where possible

File: package.json
Comment (crishpeen) [inline package.json:1]: you have engines object in this file where the yarn version is wrong

**Review by crishpeen — CHANGES_REQUESTED** [review]

File: packages/web/package.json [hunk 9]
```diff
@@ -18,9 +18,9 @@
   "scripts": {
     "design-tokens:build": "test ! -f ../design-tokens/esm/index.js && cd ../design-tokens && yarn build || true",
     "prebuild": "shx rm -rf dist && shx mkdir -p dist/scss && shx cp package.json README.md dist/ && shx cp -r src/scss dist/ && shx rm -rf dist/scss/**/__tests__",
-    "build": "npm-run-all --serial --continue-on-error design-tokens:build css js",
+    "build": "npm-run-all --serial --continue-on-error prebuild design-tokens:build css js",
```
Comment (adamkudrna) [inline packages/web/package.json:21]: As pointed out elsewhere, this could be `build:prepare` instead of `prebuild`.

```suggestion
    "build": "npm-run-all --serial --continue-on-error build:prepare design-tokens:build css js",
```

File: packages/web/package.json [hunk 10]
```diff
@@ -31,11 +31,11 @@
     "css:prefix": "postcss --config ./config/postcss.config.js --replace \"dist/css/*.css\" \"!dist/css/*.min.css\"",
```
Comment (adamkudrna) [inline packages/web/package.json:31]: `css:fix:prepare` 😄 (just kidding)

File: packages/web/package.json [hunk 11]
```diff
@@ -31,11 +31,11 @@
...
     "css:minify": "cleancss --format breaksWith=lf --source-map-inline-sources --batch --batch-suffix \".min\" \"dist/css/*.css\" \"!dist/css/*.min.css\"",
     "js": "npm-run-all js:compile",
-    "js:compile": "npm-run-all --aggregate-output --parallel js:compile:**",
+    "js:compile": "npm-run-all --aggregate-output --parallel 'js:compile:**'",
     "js:compile:standalone": "rollup --environment BUNDLE:false --config config/rollup.config.js",
     "js:compile:standalone:esm": "rollup --environment ESM:true,BUNDLE:false --config config/rollup.config.js",
     "js:compile:bundle": "rollup --environment BUNDLE:true --config config/rollup.config.js",
-    "js:compile:plugins": "tsc --module es2015 --outDir dist/js --project ./config/tsconfig.prod.json",
+    "js:compile:plugins": "tsc --module es2015 --outDir dist/js --project ./config/tsconfig.prod.json && yarn postjs:compile:plugins",
     "postjs:compile:plugins": "cd dist/js && shx mv index.esm.js index.js && shx mv index.esm.js.map index.js.map && shx mv index.esm.d.ts index.d.ts && rm index.umd.js index.umd.js.map index.umd.d.ts",
```
Comment (adamkudrna) [inline packages/web/package.json:39]: Again, this is about copying:

```suggestion
    "js:compile:plugins": "tsc --module es2015 --outDir dist/js --project ./config/tsconfig.prod.json && yarn js:copy:plugins",
    "js:copy:plugins": "cd dist/js && shx mv index.esm.js index.js && shx mv index.esm.js.map index.js.map && shx mv index.esm.d.ts index.d.ts && rm index.umd.js index.umd.js.map index.umd.d.ts",
```

**Review by adamkudrna — CHANGES_REQUESTED** [review]
Thank you for all the effort! You have done _a lot_ of work. 👍🏻 

Before merging, please revise these points:

1. Renaming the now-inlined `pre*` and `post*` tasks to something more intuitive.
2. Supporting also the short hex format, after all.

File: .yarnrc.yml (hunk 2 above)
Comment (adamkudrna) [reply .yarnrc.yml:3]: ⚠️ If there is something we developers must do to get it working with Yarn 3, please add it to the main README.

File: packages/web/package.json (hunk 10 above)
Comment (crishpeen) [reply packages/web/package.json:31]: prepare **post**css :D

File: packages/web-react/src/components/Collapse/demo/index.tsx [hunk 12]
```diff
@@ -2,18 +2,18 @@
 /* eslint-disable import/no-extraneous-dependencies, import/extensions, import/no-unresolved */
 import React from 'react';
 import ReactDOM from 'react-dom/client';
-// eslint-disable-next-line @typescript-eslint/ban-ts-comment, import/extensions, import/no-unresolved
-// @ts-ignore: No declaration file
-import icons from '@lmc-eu/spirit-icons/dist/icons';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore: No declaration file -- @see https://jira.almacareer.tech/browse/DS-561 -- @see https://jira.almacareer.tech/browse/DS-561
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-react/src/components/Collapse/demo/index.tsx]: Fixed.

File: packages/web-react/package.json [hunk 13]
```diff
@@ -74,6 +75,8 @@
     "ts-loader": "9.5.1",
     "ts-node": "10.9.2",
     "typescript": "5.3.3",
+    "webpack": "5.89.0",
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-react/package.json:78]: Yes, for building bundles, but we can remove this soon.

File: .stylelintrc.js (hunk 3 above)
Comment (literat) [reply .stylelintrc.js]: Fixed.

File: packages/design-tokens/src/scss/generated/_gradients.scss [hunk 14]
```diff
@@ -1,5 +1,5 @@
 // Generated Gradients from Supernova. Do not edit manually.
-$gradient-background-basic-overlay: linear-gradient(var(--angle, 90deg), #fff 0%, #ffffff00 100%) !default;
+$gradient-background-basic-overlay: linear-gradient(var(--angle, 90deg), #ffffff 0%, #ffffff00 100%) !default;
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/design-tokens/src/scss/generated/_gradients.scss]: Fixed.

File: packages/web-react/src/components/Header/demo/index.tsx [hunk 15]
```diff
@@ -3,16 +3,16 @@
...
+import icons from '@lmc-eu/spirit-icons/icons';
 import DocsSection from '../../../../docs/DocsSections';
 import { IconsProvider } from '../../../context';
-import HeaderMinimalInverted from './HeaderMinimalInverted';
+import HeaderExperimentalRTL from './HeaderExperimentalRTL';
 import HeaderInvertedWithActions from './HeaderInvertedWithActions';
 import HeaderInvertedWithActionsAndDialog from './HeaderInvertedWithActionsAndDialog';
+import HeaderMinimalInverted from './HeaderMinimalInverted';
 import HeaderSimpleInverted from './HeaderSimpleInverted';
 import HeaderSimpleTransparent from './HeaderSimpleTransparent';
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-react/src/components/Header/demo/index.tsx:15]: Still not part of the ESlint configuration. IDE only, for now.

File: packages/web-react/src/components/Icon/demo/index.tsx [hunk 16]
```diff
@@ -3,16 +3,17 @@
...
 import DocsSection from '../../../../docs/DocsSections';
-import Icon from '../Icon';
+import { IconsProvider } from '../../../context';
+import Icon from './Icon';
 
 ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
   <React.StrictMode>
     <IconsProvider value={icons}>
       <DocsSection title="Default">
+        {/* @ts-expect-error -- JSX element type 'Icon' does not have any construct or call signatures. */}
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-react/src/components/Icon/demo/index.tsx:16]: Just step-by-step, smaller ones :-)

File: packages/web/package.json [hunk 17]
```diff
@@ -18,9 +18,9 @@
   "scripts": {
     "design-tokens:build": "test ! -f ../design-tokens/esm/index.js && cd ../design-tokens && yarn build || true",
     "prebuild": "shx rm -rf dist && shx mkdir -p dist/scss && shx cp package.json README.md dist/ && shx cp -r src/scss dist/ && shx rm -rf dist/scss/**/__tests__",
-    "build": "npm-run-all --serial --continue-on-error design-tokens:build css js",
+    "build": "npm-run-all --serial --continue-on-error prebuild design-tokens:build css js",
     "precss": "yarn css:lint",
-    "css": "yarn css:compile && yarn css:prefix && yarn css:minify",
+    "css": "npm-run-all --serial precss css:compile css:prefix css:minify",
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web/package.json]: Fixed.

File: packages/web/package.json (hunk 10 above)
Comment (literat) [reply packages/web/package.json:30]: Happy hour! For a while, I was actually looking for needed change. Nice try, guys!

File: package.json
Comment (literat) [reply package.json:1]: Fixed.

File: packages/web-react/package.json [hunk 18]
```diff
@@ -85,7 +88,7 @@
   },
   "scripts": {
     "prebuild": "shx rm -rf dist && shx mkdir -p dist && shx cp package.json README.md dist/ ",
-    "build": "npm-run-all --serial build:umd build:es2020 build:esNext",
+    "build": "npm-run-all --serial prebuild build:umd build:es2020 build:esNext postbuild",
     "build:cjs": "yarn rollup --bundleConfigAsCjs",
     "build:umd": "npm-run-all --serial webpack:dev webpack:prod",
     "build:es2020": "tsc --module es2020 --outDir dist --project ./config/tsconfig.prod.json",
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-react/package.json:94]: I don't know if I like these jokes already...

File: packages/analytics/package.json [hunk 19]
```diff
@@ -33,8 +33,9 @@
   },
   "scripts": {
     "prebuild": "shx rm -rf dist && shx mkdir -p dist",
-    "build": "tsup --config ./config/tsup.config.ts",
+    "build": "npm-run-all --serial prebuild compile postbuild",
     "postbuild": "shx cp -r package.json README.md src/bin dist/",
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/analytics/package.json]: I have rather used `finalize` for post actions, it is more general for this usage. `cleanup` feels like I am deleting or removing something after the build. Mostly there are some copy or move actions.

File: packages/web/package.json (hunk 9 above)
Comment (literat) [reply packages/web/package.json]: Fixed.

File: packages/web/package.json (hunk 11 above)
Comment (literat) [reply packages/web/package.json]: Fixed.

File: .yarnrc.yml (hunk 2 above)
Comment (literat) [reply .yarnrc.yml]: I think it is already there. The main README file has a `Prerequisites` section with the link https://yarnpkg.com/ to the Yarn documentation. I do not think it is a good approach to copy steps from https://yarnpkg.com/getting-started/install to our docs because it can change in the future.

Information from Getting Started should be part of this PR description so you can easily migrate your development environment. But, after the merge, the prerequisites should be studied.

**Comment by literat** [comment]
@adamkudrna @crishpeen From my point of view this is read-to-go. Let's discuss the documentation of getting started with Yarn, but I think this is covered by prerequisites in the main README. Please, be short.

File: .yarn/releases/yarn-3.6.3.cjs
Comment (crishpeen) [inline .yarn/releases/yarn-3.6.3.cjs:1]: do we need to keep this version?

File: exporters/scss/package.json
Comment (crishpeen) [inline exporters/scss/package.json:1]: please update version in exporter.json, this way we can update the exporter in supernova too.

File: .yarnrc.yml [hunk 20]
```diff
@@ -0,0 +1,7 @@
+nodeLinker: node-modules
+
+plugins:
+  - path: .yarn/plugins/@yarnpkg/plugin-engines.cjs
```
Comment (crishpeen) [inline .yarnrc.yml:4]: one is local and one is remote, weird :) but I assume it is something yarn do :D

**Review by crishpeen — APPROVED** [review]
So I tested it and it works.

I just pulled the branch, `make install`, `make build` and `make start` and no problems. I didn't have to do anything more, that's great. 

So please just fix those two points I wrote earlier today and it is good to go for me. 

Good job!

File: .yarn/releases/yarn-3.6.3.cjs
Comment (literat) [reply .yarn/releases/yarn-3.6.3.cjs:1]: Bad rebase, thanks for notice. Removed.

File: exporters/scss/package.json
Comment (literat) [reply exporters/scss/package.json:1]: Updated.

File: .yarnrc.yml (hunk 20 above)
Comment (literat) [reply .yarnrc.yml:4]: Yep, the first is the local path and the second is where to download it if the file is not present. This should prevent failing when the file is missing and instead it will try to download it before it fails.

See: https://github.com/yarnpkg/berry/issues/4479

**Review by adamkudrna — APPROVED** [review]
Thank you for cleaning up the task names! 👏🏻

File: .yarnrc.yml (hunk 2 above)
Comment (adamkudrna) [reply .yarnrc.yml]: This question remains:

> Why a local instance? ~~`npm install yarn -g` is not a way to go?~~

If I understand [the docs](https://nodejs.org/api/corepack.html) correctly, Corepack should pick up the right version online when it's not available locally. But I understand this can take some time in CI.

Comment (literat) [reply .yarnrc.yml]: @crishpeen Just for clarification, the `engines` field is meant to be for the SemVer range that specifies the Node.js you should use for the package to work with. Using this for defining the package manager can cause some trouble.
